src/overlays/actors/ovl_Obj_Mine/z_obj_mine.c

```diff
@@ -24,131 +53,1182 @@ ActorInit Obj_Mine_InitVars = {
+        AC_ON | AC_TYPE_PLAYER,
+        OC1_ON | OC1_TYPE_PLAYER | OC1_TYPE_1,
+        OC2_TYPE_1,
+        COLSHAPE_JNTSPH,
+    },
+    ARRAY_COUNT(sJntSphElementsInit),
+    sJntSphElementsInit,
+};
+
+static f32 sPathSpeeds[] = {
+    1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f, 7.0f, 8.0f,
 };
 
-// static InitChainEntry sInitChain[] = {
-static InitChainEntry D_80A845E8[] = {
+static ObjMineMtxF3 sStandardBasis = {
+    { 1.0f, 0.0f, 0.0f },
+    { 0.0f, 1.0f, 0.0f },
+    { 0.0f, 0.0f, 1.0f },
+};
+
+static InitChainEntry sInitChain[] = {
     ICHAIN_F32(uncullZoneForward, 1300, ICHAIN_CONTINUE),
     ICHAIN_F32(uncullZoneScale, 150, ICHAIN_CONTINUE),
     ICHAIN_F32(uncullZoneDownward, 100, ICHAIN_CONTINUE),
     ICHAIN_VEC3F_DIV1000(scale, 10, ICHAIN_STOP),
 };
 
-#endif
+void ObjMine_Path_MoveToWaypoint(ObjMine* this, s32 index) {
+    Math_Vec3s_ToVec3f(&this->actor.world.pos, &this->waypoints[index]);
+}
+
+s32 ObjMine_GetUnitVec3f(Vec3f* src, Vec3f* dst) {
+    f32 magnitude = Math3D_Vec3fMagnitude(src);
+
+    if (magnitude < 0.001f) {
+        return false;
+    }
+
+    dst->x = src->x * (1.0f / magnitude);
+    dst->y = src->y * (1.0f / magnitude);
+    dst->z = src->z * (1.0f / magnitude);
+    return true;
+}
+
+s32 ObjMine_GetUnitVec3fNorm(Vec3f* src, Vec3f* dst, f32* norm, f32* invNorm) {
+    f32 magnitude = Math3D_Vec3fMagnitude(src);
+
+    if (magnitude < 0.001f) {
+        return false;
+    }
+
+    dst->x = src->x * (1.0f / magnitude);
+    dst->y = src->y * (1.0f / magnitude);
+    dst->z = src->z * (1.0f / magnitude);
+    *norm = magnitude;
+    *invNorm = 1.0f / magnitude;
+
+    return true;
+}
+
+void ObjMine_Path_SpawnBomb(ObjMine* this, PlayState* play) {
+    EnBom* bomb = (EnBom*)Actor_Spawn(&play->actorCtx, play, ACTOR_EN_BOM, this->actor.world.pos.x,
+                                      this->actor.world.pos.y - BOMB_SPAWN_OFFSET, this->actor.world.pos.z, 0, 0, 0,
+                                      BOMB_TYPE_BODY);
+
+    if (bomb != NULL) {
+        bomb->timer = 0;
+    }
+}
+
+void ObjMine_AirWater_SpawnBomb(ObjMine* this, PlayState* play) {
+    f32 x = this->collider.elements[0].dim.worldSphere.center.x;
+    f32 y = this->collider.elements[0].dim.worldSphere.center.y - BOMB_SPAWN_OFFSET;
+    f32 z = this->collider.elements[0].dim.worldSphere.center.z;
+    EnBom* bomb = (EnBom*)Actor_Spawn(&play->actorCtx, play, ACTOR_EN_BOM, x, y, z, 0, 0, 0, BOMB_TYPE_BODY);
+
+    if (bomb != NULL) {
+        bomb->timer = 0;
+    }
+}
+
+s32 ObjMine_AirWater_CheckOC(ObjMine* this) {
+    if (this->collider.base.ocFlags2 & OC2_HIT_PLAYER) {
+        return true;
+    }
+
+    if (this->collider.base.ocFlags1 & OC1_HIT) {
+        Actor* hitActor = this->collider.base.oc;
+
+        if ((hitActor->id == ACTOR_OBJ_MINE) && (hitActor->room == this->actor.room)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void ObjMine_Air_CheckAC(ObjMine* this, s16* hitAngle, s16* torqueAngle) {
```

Comment:
I think these ones that have other side effects like calculating knockback directions and so on should have some function comments to that effect, since it's not clear just from the name.